core/src/main/java/org/geysermc/geyser/session/cache/SkullCache.java

```diff
@@ -0,0 +1,209 @@
+        } else {
+            if (!cullingEnabled) {
+                assignSkullEntity(skull);
+                return;
+            }
+            if (lastPlayerPosition == null) {
+                return;
+            }
+            skull.distanceSquared = position.distanceSquared(lastPlayerPosition.getX(), lastPlayerPosition.getY(), lastPlayerPosition.getZ());
+            if (skull.distanceSquared < skullRenderDistanceSquared) {
+                // Keep list in order
+                int i = Collections.binarySearch(inRangeSkulls, skull, Comparator.comparingInt(Skull::getDistanceSquared));
+                if (i < 0) { // skull.distanceSquared is a new distance value
+                    i = -i - 1;
+                }
+                inRangeSkulls.add(i, skull);
+
+                if (i < maxVisibleSkulls) {
+                    // Reassign entity from the farthest skull to this one
+                    if (inRangeSkulls.size() > maxVisibleSkulls) {
+                        freeSkullEntity(inRangeSkulls.get(maxVisibleSkulls));
+                    }
+                    assignSkullEntity(skull);
+                }
+            }
+        }
+    }
+
+    public void removeSkull(Vector3i position) {
+        Skull skull = skulls.remove(position);
+        if (skull != null && skull.entity != null) {
+            freeSkullEntity(skull);
+
+            if (cullingEnabled) {
+                inRangeSkulls.remove(skull);
+                if (inRangeSkulls.size() >= maxVisibleSkulls) {
+                    // Reassign entity to the closest skull without an entity
+                    assignSkullEntity(inRangeSkulls.get(maxVisibleSkulls - 1));
+                }
+            }
+        }
+    }
+
+    public void updateVisibleSkulls() {
+        if (cullingEnabled) {
+            // No need to recheck skull visibility for small movements
+            if (lastPlayerPosition != null && session.getPlayerEntity().getPosition().distanceSquared(lastPlayerPosition) < 4) {
+                return;
+            }
+            lastPlayerPosition = session.getPlayerEntity().getPosition();
+            inRangeSkulls.clear();
+            for (Skull skull : skulls.values()) {
+                skull.distanceSquared = skull.position.distanceSquared(lastPlayerPosition.getX(), lastPlayerPosition.getY(), lastPlayerPosition.getZ());
+                if (skull.distanceSquared > skullRenderDistanceSquared) {
+                    freeSkullEntity(skull);
+                } else {
+                    inRangeSkulls.add(skull);
+                }
+            }
+            inRangeSkulls.sort(Comparator.comparingInt(Skull::getDistanceSquared));
+
+            for (int i = maxVisibleSkulls; i < inRangeSkulls.size(); i++) {
+                freeSkullEntity(inRangeSkulls.get(i));
+            }
+            for (int i = 0; i < Math.min(maxVisibleSkulls, inRangeSkulls.size()); i++) {
+                assignSkullEntity(inRangeSkulls.get(i));
+            }
```

Comment:
Shouldn't it be possible to combine these 2 since it's a sorted list? It would prevent unneeded movement and metadata packets.
We could just check if `i < maxVisibleSkulls`. When it returns true, we run the `assignSkullEntity` method (which automatically checks if a skull already has an entity). When it returns false, we run the `freeSkullEntity` method.